src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/PartiallyDeterminedHaplotypeComputationEngine.java

```diff
@@ -95,152 +95,91 @@ public static AssemblyResultSet generatePDHaplotypes(final AssemblyResultSet sou
+                                .sorted(HAPLOTYPE_SNP_FIRST_COMPARATOR).toList();
+                        PartiallyDeterminedHaplotype newPDHaplotype = createNewPDHaplotypeFromEvents(referenceHaplotype, determinedEvents, determinedLocus, eventsInPDHap, allEventsHere);
+                        branchHaplotypesDebugMessage(referenceHaplotype, debug, branch, List.of(newPDHaplotype));
+                        outputHaplotypes.add(newPDHaplotype);
                     } else {
                         // TODO currently this approach doesn't properly handle a bunch of duplicate events...
-                        // If we are producing determined bases, then we want to enforce that every new event at least has THIS event as a variant.
-                        List<List<Event>> growingEventGroups = new ArrayList<>();
-                        growingEventGroups.add(new ArrayList<>());
-                        for (final int otherEventGroupStart : variantsByStartPos.keySet()) {
-                            // Iterate through the growing combinatorial expansion of haps, split it into either having or not having the variant.
-                            if (growingEventGroups.size() > MAX_BRANCH_PD_HAPS) {
-                                Utils.printIf(debug, () -> "Too many branch haplotypes ["+growingEventGroups.size()+"] generated from site, falling back on assembly variants!");
-                                return sourceSet;
-                            } else if (otherEventGroupStart == thisEventGroupStart) {
-                                growingEventGroups.forEach(group -> group.add(determinedEventToTest));
-                            } else if (thisEventGroupStart < otherEventGroupStart) {
-                                variantsByStartPos.get(otherEventGroupStart).stream()
-                                        .filter(event -> !excludeEvents.contains(event))
-                                        .flatMap(event -> growingEventGroups.stream().map(group -> growEventGroup(group, event)))
-                                        .forEach(growingEventGroups::add);
+                        // Start with a single "seed" haplotype containing the determined event(s). At each downstream locus,
+                        // every haplotype can grow by every event, or remain unchanged.  For example, if we have haplotypes
+                        // H1 and H2 so far and reach a locus with events A and B our list grows to H1/ref, H2/ref, H1/A, H1/B, H2/A, H2/B.
+                        List<List<Event>> haplotypeEvents = new ArrayList<>();
+                        haplotypeEvents.add(new ArrayList<>(determinedEvents)); // the seed haplotype
+
+                        for (final int locus : eventsByStartPos.keySet()) {
+                            if (determinedLocus < locus) {
+                                final List<List<Event>> children = eventsByStartPos.get(locus).stream().filter(branch::contains)
```

Comment:
this is already within a branch right? and a branch is now a positive list of events that might exist? It seems concevable that you could simply iteraate off of the branch here.... This whole codepath is less tested and I would make sure there is actually an integration test for this branching paths code... I'm not sure i trust this refactor without it having an integration test in the HC integration tests.

Reply:
I wouldn't worry about the integration tests.  They caught all of my mistakes while working on this PR, and I double-checked on a whole genome.